src/hotspot/share/opto/inlinetypenode.cpp

```diff
@@ -631,6 +631,158 @@ void InlineTypeNode::store(GraphKit* kit, Node* base, Node* ptr, bool immutable_
   }
 }
 
+// Adds a check between val1 and val2. Jumps to 'region' if check passes and optionally sets the corresponding phi input to false.
+static void acmp_val_guard(PhaseIterGVN* igvn, RegionNode* region, Node* phi, Node** ctrl, BasicType bt, BoolTest::mask test, Node* val1, Node* val2) {
+  Node* cmp = nullptr;
+  switch (bt) {
+  case T_FLOAT:
+    val1 = igvn->register_new_node_with_optimizer(new MoveF2INode(val1));
+    val2 = igvn->register_new_node_with_optimizer(new MoveF2INode(val2));
+    // Fall-through to the int case
+  case T_BOOLEAN:
+  case T_CHAR:
+  case T_BYTE:
+  case T_SHORT:
+  case T_INT:
+    cmp = igvn->register_new_node_with_optimizer(new CmpINode(val1, val2));
+    break;
+  case T_DOUBLE:
+    val1 = igvn->register_new_node_with_optimizer(new MoveD2LNode(val1));
+    val2 = igvn->register_new_node_with_optimizer(new MoveD2LNode(val2));
+    // Fall-through to the long case
+  case T_LONG:
+    cmp = igvn->register_new_node_with_optimizer(new CmpLNode(val1, val2));
+    break;
+  default:
+    assert(is_reference_type(bt), "must be");
+    cmp = igvn->register_new_node_with_optimizer(new CmpPNode(val1, val2));
+  }
+  Node* bol = igvn->register_new_node_with_optimizer(new BoolNode(cmp, test));
+  IfNode* iff = igvn->register_new_node_with_optimizer(new IfNode(*ctrl, bol, PROB_MAX, COUNT_UNKNOWN))->as_If();
+  Node* if_f = igvn->register_new_node_with_optimizer(new IfFalseNode(iff));
+  Node* if_t = igvn->register_new_node_with_optimizer(new IfTrueNode(iff));
+
+  region->add_req(if_t);
+  if (phi != nullptr) {
+    phi->add_req(igvn->intcon(0));
+  }
+  *ctrl = if_f;
+}
+
+// Check if a substitutability check between 'this' and 'other' can be implemented in IR
+bool InlineTypeNode::can_emit_substitutability_check(Node* other) const {
+  if (other != nullptr && other->is_InlineType() && bottom_type() != other->bottom_type()) {
+    // Different types, this is dead code because there's a check above that guarantees this.
```

Comment:
Shouldn't this be an assert, then? Or at least call it a sanity check instead, just so no one gets an idea and removes it.